Api(client): 대시보드 api 연결

karnelll opened the pull request:
## 📌 Related Issues

> 관련된 Issue를 태그해주세요. (e.g. - close #25)

- close #100

## ✅ 체크 리스트

- [x] PR 제목의 형식을 잘 작성했나요? e.g. Feat(client): PR 템플릿 작성
- [ ] 빌드가 성공했나요? (pnpm build)

## 📄 Tasks

<!-- 해당 PR에 대한 작업 내용을 요약하여 작성해주세요. (없을 경우 section 삭제) -->

### 🔧 대시보드 API 연동 및 타입 구조 리팩토링

**API 연동 작업:**
- **사용자 API**: `userApi.ts` - 도토리 개수 및 리마인더 시간 조회
- **아티클 API**: `articleApi.ts` - 아티클 목록, 상세, 읽음 상태 업데이트
- **카테고리 API**: `categoryApi.ts` - 카테고리별 아티클 조회
- **썸네일 API**: `thumbnail.ts` - 아티클 썸네일 생성 및 조회

**주요 API 기능:**
- `getAcornCountWithRemind()`: 사용자 도토리 개수 및 다음 리마인더 시간 조회
- `getArticlesByCategory()`: 카테고리별 아티클 목록 조회 (페이지네이션 지원)
- `getUnreadArticles()`: 읽지 않은 아티클 목록 조회
- `getDailyReminderArticles()`: 데일리 리마인더 아티클 조회
- `updateArticleReadStatus()`: 아티클 읽음 상태 업데이트 (도토리 수집)
- `createTestArticles()`: 테스트용 아티클 생성

### 썸네일 API 연동 상세

**썸네일 생성 로직:**
```typescript
generateThumbnail(url: string): Promise<string | null>
```

**썸네일 추출 우선순위:**
1. **Open Graph 이미지** (`og:image` 메타 태그)
2. **Twitter 이미지** (`twitter:image` 메타 태그)
3. **Favicon** (Google Favicon 서비스 활용)
4. **기본 이미지** (추출 실패 시)

**기술적 구현:**
- **프록시 서비스**: `https://api.allorigins.win/raw` 사용하여 CORS 우회
- **Favicon 서비스**: `https://www.google.com/s2/favicons` 활용
- **정규식 패턴**: 메타 태그에서 이미지 URL 추출
- **URL 정규화**: 상대 경로를 절대 경로로 변환

**에러 처리:**
- 프록시 서비스 실패 시 favicon으로 fallback
- 이미지 추출 실패 시 null 반환하여 기본 이미지 표시
- 네트워크 오류 시 적절한 로깅 및 에러 핸들링

**성능 최적화:**
- 도메인 추출을 통한 효율적인 favicon URL 생성
- 메타데이터 우선 확인 후 favicon 사용
- 불필요한 네트워크 요청 최소화

**타입 구조 개선:**
- `apps/client/src/pages/dashboard/apis/api.ts` → `apps/client/src/pages/dashboard/types/api.ts`로 타입 파일 이동
- 모든 API 관련 타입들을 `types` 폴더로 통합하여 구조 개선
- 대시보드 내 모든 컴포넌트 및 API 파일의 import 경로 일괄 수정

**수정된 파일:**
- `articleApi.ts`, `userApi.ts`, `categoryApi.ts` - API 타입 import 경로 수정
- `BookmarkSection.tsx`, `DailyReminderSection.tsx` - 컴포넌트 타입 import 경로 수정
- `useDashboard.ts`, `dashboard.tsx` - 훅 및 메인 컴포넌트 타입 import 경로 수정

**개선 효과:**
- 대시보드 페이지의 완전한 API 연동으로 실제 데이터 연동
- 아티클 URL에 대한 자동 썸네일 생성으로 시각적 개선
- 타입 정의의 일관성 있는 관리
- 코드 구조의 명확성 향상
- 향후 타입 관리 용이성 증대

## ⭐ PR Point (To Reviewer)

<!-- 리뷰어에게 추가로 전달할 사항이 있다면 작성해주세요. (없을 경우 section 삭제) -->

### 리뷰 포인트

1. **API 연동 완성도**: 모든 대시보드 기능이 실제 API와 연동되었는지 확인
2. **썸네일 생성 로직**: 다양한 웹사이트에서 썸네일이 올바르게 추출되는지 확인
3. **에러 처리**: API 호출 시 적절한 에러 처리가 구현되었는지 확인
4. **타입 구조 변경**: API 타입들을 `types` 폴더로 이동한 것이 적절한지 확인
5. **Import 경로 수정**: 모든 파일에서 올바른 경로로 수정되었는지 확인
6. **기능 영향도**: 타입 이동으로 인한 기능상 문제가 없는지 확인

### 추가 작업 사항

- 기존 `@shared/types`에서 대시보드 전용 타입들을 분리하여 대시보드 독립성 확보
- API 응답 타입과 UI 컴포넌트 타입을 명확히 분리
- 향후 다른 페이지에서도 동일한 패턴으로 타입 구조 정리 예정

### 썸네일 API 주요 특징

- **다단계 추출**: OG 이미지 → Twitter 이미지 → Favicon 순서로 우선순위 적용
- **CORS 우회**: 프록시 서비스를 통한 안전한 메타데이터 추출
- **Fallback 메커니즘**: 각 단계별 실패 시 다음 단계로 자동 전환
- **URL 정규화**: 상대 경로를 절대 경로로 자동 변환
- **성능 최적화**: 불필요한 네트워크 요청 최소화

### API 연동 주요 특징

- **페이지네이션 지원**: 아티클 목록 조회 시 페이지 단위 로딩
- **실시간 상태 업데이트**: 아티클 읽음 상태 변경 시 도토리 개수 실시간 반영
- **카테고리별 필터링**: 사용자가 선택한 카테고리별 아티클 조회
- **데일리 리마인더**: 설정된 시간에 맞춰 리마인더 아티클 제공

## 📷 Screenshot

<!-- 작업한 내용에 대한 자료가 필요하다면 첨부해주세요. (없을 경우 section 삭제)-->

**변경 전:**
```
apis/
├── api.ts (타입 정의 포함)
├── articleApi.ts
└── userApi.ts
```

**변경 후:**
```
types/
├── api.ts (API 타입 정의)
└── components.ts (UI 컴포넌트 타입)

apis/
├── articleApi.ts (아티클 API 연동)
├── userApi.ts (사용자 API 연동)
├── categoryApi.ts (카테고리 API 연동)
└── thumbnail.ts (썸네일 API 연동)
```

**썸네일 생성 플로우:**
```
URL 입력
    ↓
프록시 서비스로 메타데이터 요청
    ↓
OG 이미지 추출 시도
    ↓ (실패 시)
Twitter 이미지 추출 시도
    ↓ (실패 시)
Favicon 생성
    ↓ (실패 시)
기본 이미지 표시
```

**API 연동 구조:**
```
Dashboard Components
    ↓
useDashboard Hook
    ↓
API Functions (articleApi, userApi, categoryApi, thumbnail)
    ↓
Backend API Endpoints
```


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

**새로운 기능**
- 대시보드가 기사, 카테고리, 도토리 개수 등 모든 데이터를 실시간 API로 불러오도록 전면 개선되었습니다.
- 기사, 북마크, 리마인더 카드에 썸네일 자동 생성, 로딩 및 에러 처리 기능이 추가되었습니다.
- 기사 읽음 처리, 카테고리별 필터링, 읽지 않은 기사 수 표시 등 사용자 인터랙션이 강화되었습니다.
- 로딩 상태, 스켈레톤 UI, 비어있는 상태 등 다양한 사용자 경험 개선이 적용되었습니다.
- 대시보드 관련 API, 타입, 훅, 유틸리티가 새롭게 도입되어 기능 분리 및 관리가 용이해졌습니다.

**버그 수정**
- 인증 토큰 만료 또는 부재 시 온보딩 페이지로 안전하게 리다이렉트되도록 처리되었습니다.
- 헤더가 최상단에 고정되어 다른 요소와 겹치는 현상이 해결되었습니다.

**리팩터링 및 구조 개선**
- 대시보드 데이터 관리를 위해 여러 커스텀 훅으로 분리 및 통합하여 유지보수성과 확장성이 향상되었습니다.
- 불필요한 목데이터, 사용하지 않는 API, 타입, 훅 파일이 대거 정리되었습니다.
- 내부 경로 및 타입 전용 import 적용 등 코드 일관성이 개선되었습니다.

**스타일**
- 카드, 버튼, 배너 등 주요 UI 컴포넌트 스타일과 인터페이스가 통일되고 개선되었습니다.

**문서화 및 타입**
- 대시보드 API 응답 및 컴포넌트 타입이 명확하게 정의되어 타입 안정성이 강화되었습니다.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Commit:
fix: 썸네일 캐싱 및 에러 처리 개선, 환경변수 적용

Commit:
refactor: 썸네일 API를 대시보드 전용 폴더로 이동 및 구조 정리

Commit:
fix: ESLint 오류 수정 및 코드 정리

- 빈 함수 오류 수정 (empty arrow function)
- non-null assertion 오류 수정
- 디버깅용 console.log 제거
- 코드 포맷팅 적용

Commit:
feat: 대시보드 API 연동

# Conflicts:
#	apps/client/src/pages/dashboard/dashboard.tsx
#	apps/client/src/pages/onBoarding/OnBoarding.tsx

Commit:
refactor: api 타입 분리 및 import 경로 일괄 수정

# Conflicts:
#	apps/client/src/pages/dashboard/dashboard.tsx

Commit:
fix: 이메일 하드코딩 환경변수 처리 및 fallback 적용

# Conflicts:
#	apps/client/src/pages/dashboard/dashboard.tsx

Commit:
refactor: 대시보드 코드 정리 및 최적화

- TestSection 컴포넌트 삭제
- 테스트 버튼 UI 제거
- useDashboard 훅에서 테스트 관련 함수 제거
- 도토리 개수 업데이트 로직 개선
- 배너 렌더링 디버깅 로그 추가
- 썸네일 화질 개선 (64px → 128px)
- api.allorigins.win 프록시 서비스 제거
- TypeScript 오류 수정

# Conflicts:
#	apps/client/src/pages/dashboard/dashboard.tsx

Commit:
fix: 카드 더보기 버튼 이벤트 중첩 해결 및 토큰 관리 개선

- BookmarkCard, DailyReminderCard에서 더보기 버튼 클릭 시 e.stopPropagation() 추가
- DailyReminderCard 스탬프 오버레이에 pointer-events-none 추가
- Dashboard 모달 팝업 시 스크롤 고정 로직 추가
- axiosInstance에서 하드코딩된 토큰 발급 로직 제거
- 온보딩에서 저장된 토큰만 사용하도록 수정
- 대시보드에서 토큰이 없으면 온보딩으로 리다이렉트하는 로직 추가

Commit:
fix: 대시보드 토큰 통합 및 이벤트 처리 문제 수정

constantly-dev commented on apps/client/src/pages/dashboard/apis/articleApi.ts:
```diff
@@ -0,0 +1,106 @@
+    return data;
+  } catch (error) {
+    logApiError('getArticleDetail', error);
+    throw error;
```
테스트 용으로 사용하신 log는 지워주세요!

karnelll replied:
네!!

constantly-dev commented on apps/client/src/pages/dashboard/apis/articleApi.ts:
```diff
@@ -0,0 +1,106 @@
+
+export const getArticleDetail = async (
+  articleId: number
+): Promise<ApiResponse<ArticleDetail>> => {
```
`ApiResponse` 커스텀 잘 해주셨네요! 다만 Axios에서 `AxiosResponse`라는 타입을 제공해주기도 합니다! 참고하면 좋을 것 같아요!

karnelll replied:
제공하는 타입 활용해서 수정하겠습니다!

constantly-dev commented on apps/client/src/pages/dashboard/apis/articleApi.ts:
```diff
@@ -0,0 +1,106 @@
+    );
+    logApiSuccess('getArticlesByCategory', data);
+    return data;
+  } catch (error: any) {
```
`error`도 `axiosError`와 같은 타입으로 구체화하면 좋을 것 같아요!

karnelll replied:
넵!!

Commit:
fix: 콘솔 로그 및 errorUtils 파일 제거 및 개발자 테스트용 API 및 관련 코드 제거

constantly-dev commented on apps/client/src/pages/dashboard/apis/thumbnail.ts:
```diff
@@ -0,0 +1,185 @@
+  try {
+    const response = await fetch(url, {
+      signal: controller.signal,
+      headers: { 'Content-Type': 'text/html' },
```
해당 headers가 필요한 것인가요??

Commit:
fix: api 폴더에서 썸네일 유틸리티를 utils 폴더로 분리

Commit:
fix: TypeScript 에러 수정

Commit:
fix: 주석 정리 및 간소화

Commit:
fix: axiosInstance에서 온보딩 토큰 사용하도록 수정

Commit:
fix: 주석 지우기

Commit:
fix: 주석 정리

constantly-dev commented on apps/client/src/pages/onBoarding/components/TimeSelectStep.tsx:
```diff
@@ -53,7 +53,7 @@ const TimeSelectStep = ({ setStep, os, email }: TimeSelectStepProps) => {
           {
             onSuccess: (response) => {
-              onSigninSuccess(response.data.token);
+              onSigninSuccess(response.data.token, email || undefined);
```
제가 email 빼먹었네요..!!! 굿굿 👍

constantly-dev reviewed (dismissed):
수고하셨습니다~~ 👍

Commit:
refactor: API 호출 함수 정리 및 썸네일 유틸리티 분리

- API 호출 함수에서 불필요한 try/catch 및 콘솔 로그 제거
- 썸네일 관련 유틸리티를 apis에서 utils로 분리
- Thumbnail 컴포넌트 import 경로 수정
- useAcornCount 훅에서 불필요한 try/catch 제거

Commit:
refactor: AxiosResponse 타입 활용 및 코드 개선

- API 함수들이 AxiosResponse 타입을 반환하도록 수정
- 훅들에서 response.data로 접근하도록 수정
- 사용하지 않는 AxiosError import 제거
- 썸네일 유틸리티에서 불필요한 Content-Type 헤더 제거

constantly-dev reviewed (approved):
👍
